### PR TITLE
ENH added code to rotate anaconda tokens

### PR DESCRIPTION
**Pull request by beckermr**
This PR is cleaned up version of the token rotation code from #1283. I will rebase that PR on master once this one goes in.

Checklist
* [x] Added a ``news`` entry

### CODE REVIEW

**Comment by beckermr** [comment]
OK @conda-forge/core @isuruf @scopatz This one is ready to be reviewed and merged!

**Comment by beckermr** [comment]
@conda-forge/core can someone check this over? It is needed for CFEP-13.

**Comment by isuruf** [comment]
Can we use another name? Otherwise there'll be disruption between the time the token is changed and there's a rerender.

**Comment by beckermr** [comment]
Another name for what? The token on disk?

**Comment by isuruf** [comment]
Name of the token. (`BINSTAR_TOKEN`)

**Comment by beckermr** [comment]
I'm confused about the issue here. Can you be more specific? Using this command does not require a rerender except for appveyor.

**Comment by isuruf** [comment]
If I'm understanding this correctly, this PR is used for changing the BINSTAR_TOKEN for `conda-forge` account into `cf-staging`. AFAIK, to use `cf-staging`, you need a rerender, right?

**Comment by beckermr** [comment]
Ahhhh so I think we might be conflating two different things here. On the one hand, we could use the code here to rotate the binstar tokens we have. This utility seems quite general. 

On the other, I do plan to use this for cfep-13. My plan was to rerender each feedstock and push straight to master without a rebuild. A full rebuild would be expensive and seems unnecessary. I know this procedure wouldn’t be great for reproducibility, but rebuilding every feedstock in conda forge is a big task and not needed. If we are concerned about reproducibility we could add a git tag to each feedstock before we push so that we can mark the last build before the change. 

Even if we don’t end up with the plan above, the code in this pr is still quite general and so could be merged as is.

**Comment by beckermr** [comment]
Btw, if we do end up wanting a separate token name, then we can push a change later to have the token name be configurable. I could add that to this pr now if you’d like too.

**Comment by isuruf** [comment]
> My plan was to rerender each feedstock and push straight to master without a rebuild. 

That's too invasive. This would make all PRs stale and some feedstocks have some custom tweaks that they reapply every rerender.

> Btw, if we do end up wanting a separate token name, then we can push a change later to have the token name be configurable. I could add that to this pr now if you’d like too.

Sure.

**Comment by isuruf** [comment]
Also, I'd like to have feedstock master working and reproducible. A forced rerender might bring in changes that make the feedstock not work anymore.

**Comment by beckermr** [comment]
> That's too invasive. This would make all PRs stale and some feedstocks have some custom tweaks that they reapply every rerender.

I understand the concern about that for sure, but all PRs will be stale eventually anyways. We have to turn off the old token at some point. Further, even if we issued a rebuild PR to every feedstock and a large fraction of them were merged, we'd still leave the current PRs on the feedstocks stale until they were rerendered (if they had merged the CFEP-13 rebuild or not). I don't see the actual difference between the procedure I suggested above and the alternative and making true rebuild PRs with separate tokens.

**Comment by beckermr** [comment]
The downside of full rebuild PRs is that they are quite taxing on our volunteer maintainers.

**Comment by beckermr** [comment]
> Also, I'd like to have feedstock master working and reproducible. A forced rerender might bring in changes that make the feedstock not work anymore.

I agree. We can apply a git tag so we can recover the feedstock as it was.

**Comment by isuruf** [comment]
My suggestion was not to make full rebuild PRs. Sorry for not making it clear.
My suggestion is to add both `BINSTAR_TOKEN` and `STAGING_BINSTAR_TOKEN` to `conda-forge.yml` and when there's a new rerender by a user with the new conda-smithy, it will rerender with `STAGING_BINSTAR_TOKEN`. After a few weeks then we remove `BINSTAR_TOKEN`.

**Comment by beckermr** [comment]
> My suggestion was not to make full rebuild PRs. Sorry for not making it clear.
My suggestion is to add both BINSTAR_TOKEN and STAGING_BINSTAR_TOKEN to conda-forge.yml and when there's a new rerender by a user with the new conda-smithy, it will rerender with STAGING_BINSTAR_TOKEN. After a few weeks then we remove BINSTAR_TOKEN.

Sure we can do that, but most feedstocks don't see regular rebuilds. So even if we have an in-between state, we eventually leave a bunch of PRs and feedstocks behind.

**Comment by isuruf** [comment]
True. My main concern is for PRs already with a green button which were rerendered with an old version. If we force a rerender on master (which is not good for reproducability), these would become stale. If we don't, after merging these would fail. That's why I suggested a compromise.

**Comment by beckermr** [comment]
Sounds good. I’ll need to push an adjustment to the ci setup package to use the staging token if validating.

File: conda_smithy/feedstock_tokens.py [hunk 1]
```diff
@@ -618,6 +618,8 @@ def add_feedstock_token_to_azure(user, project, feedstock_token, clobber):
             "Cannot add FEEDSTOCK_TOKEN to a repo that is not already registerd on azure CI!"
         )
 
+    ed = bclient.get_definition(ed.id, project=config.project_name)
```
Comment (beckermr) [inline conda_smithy/feedstock_tokens.py:621]: This was a bug I found.

**Comment by beckermr** [comment]
OK @isuruf. I put in the requested changes. I fixed three bugs.

1. The code now checks the conda-forge.yml to make sure any travis tokens there with the same name are gone. These tokens will override the ones set in the API. It also warns the the feedstock will need to be rerendered in certain cases when updating tokens in travis.
2. I fixed a bug in the feedstock token registration that deleted other secrets on azure.
3. I fixed the argparse calls so that the output of the subcommands is correct.

This one is ready for another look!

**Comment by beckermr** [comment]
thanks!

**Pull request merged**